test/mocks/network/connection.h

```diff
@@ -45,48 +45,61 @@ class MockConnectionBase {
+  MOCK_METHOD(void, setDelayedCloseTimeout, (std::chrono::milliseconds));                          \
+  MOCK_METHOD(absl::string_view, transportFailureReason, (), (const));                             \
+  MOCK_METHOD(absl::optional<std::chrono::milliseconds>, lastRoundTripTime, (), (const))
+
 class MockConnection : public Connection, public MockConnectionBase {
 public:
   MockConnection();
   ~MockConnection() override;
+  DEFINE_MOCK_CONNECTION_MOCK_METHODS;
+};
 
-  // Network::Connection
-  MOCK_METHOD(void, addConnectionCallbacks, (ConnectionCallbacks & cb));
-  MOCK_METHOD(void, addBytesSentCallback, (BytesSentCb cb));
-  MOCK_METHOD(void, addWriteFilter, (WriteFilterSharedPtr filter));
-  MOCK_METHOD(void, addFilter, (FilterSharedPtr filter));
-  MOCK_METHOD(void, addReadFilter, (ReadFilterSharedPtr filter));
-  MOCK_METHOD(void, enableHalfClose, (bool enabled));
-  MOCK_METHOD(void, close, (ConnectionCloseType type));
-  MOCK_METHOD(Event::Dispatcher&, dispatcher, ());
-  MOCK_METHOD(uint64_t, id, (), (const));
-  MOCK_METHOD(void, hashKey, (std::vector<uint8_t>&), (const));
-  MOCK_METHOD(bool, initializeReadFilters, ());
-  MOCK_METHOD(std::string, nextProtocol, (), (const));
-  MOCK_METHOD(void, noDelay, (bool enable));
-  MOCK_METHOD(void, readDisable, (bool disable));
-  MOCK_METHOD(void, detectEarlyCloseWhenReadDisabled, (bool));
-  MOCK_METHOD(bool, readEnabled, (), (const));
-  MOCK_METHOD(const Address::InstanceConstSharedPtr&, remoteAddress, (), (const));
-  MOCK_METHOD(const Address::InstanceConstSharedPtr&, directRemoteAddress, (), (const));
-  MOCK_METHOD(absl::optional<Connection::UnixDomainSocketPeerCredentials>,
-              unixSocketPeerCredentials, (), (const));
-  MOCK_METHOD(const Address::InstanceConstSharedPtr&, localAddress, (), (const));
-  MOCK_METHOD(void, setConnectionStats, (const ConnectionStats& stats));
-  MOCK_METHOD(Ssl::ConnectionInfoConstSharedPtr, ssl, (), (const));
-  MOCK_METHOD(absl::string_view, requestedServerName, (), (const));
-  MOCK_METHOD(State, state, (), (const));
-  MOCK_METHOD(void, write, (Buffer::Instance & data, bool end_stream));
-  MOCK_METHOD(void, setBufferLimits, (uint32_t limit));
-  MOCK_METHOD(uint32_t, bufferLimit, (), (const));
-  MOCK_METHOD(bool, localAddressRestored, (), (const));
-  MOCK_METHOD(bool, aboveHighWatermark, (), (const));
-  MOCK_METHOD(const Network::ConnectionSocket::OptionsSharedPtr&, socketOptions, (), (const));
-  MOCK_METHOD(StreamInfo::StreamInfo&, streamInfo, ());
-  MOCK_METHOD(const StreamInfo::StreamInfo&, streamInfo, (), (const));
-  MOCK_METHOD(void, setDelayedCloseTimeout, (std::chrono::milliseconds));
-  MOCK_METHOD(absl::string_view, transportFailureReason, (), (const));
-  MOCK_METHOD(absl::optional<std::chrono::milliseconds>, lastRoundTripTime, (), (const));
+class MockServerConnection : public ServerConnection, public MockConnectionBase {
```

Comment:
Would it make sense to mimic the real ConnectionImpl and use "public virtual ServerConnection" to get this stuff to work?  Or a template<typename BaseT> MockConnectionTmpl : public BaseT that can provides the base implementations of most mock methods?

class MockConnection : public MockConnectionTmpl<Connection>, public MockConnectionBase {
   ...
};

Reply:
Supposedly this is just an issue with gMock. From [the docs](https://github.com/google/googletest/blob/d11c76175f24f23a1fd316fe37d3743133479e27/googlemock/docs/cook_book.md#the-nice-the-strict-and-the-naggy-nicestrictnaggy):

> There are some caveats though (sadly they are side effects of C++'s limitations):

> NiceMock<MockFoo> and StrictMock<MockFoo> only work for mock methods defined using the MOCK_METHOD macro directly in the MockFoo class. If a mock method is defined in a base class of MockFoo, the “nice” or “strict” modifier may not affect it, depending on the compiler.

Otherwise virtual inheritance would be the right thing to do.